### PR TITLE
First attempt at ReLU

### PR DESCRIPTION
**Pull request by plavin**
I'm getting a segfault when I run this, haven't figured out why yet. I'll look more later. 

Error:

```
libc++abi.dylib: terminating with uncaught exception of type af::exception: ArrayFire Exception (Input types are not the same:205):
In function af_err af_matmul(af_array *, const af_array, const af_array, const af_mat_prop, const af_mat_prop)
In file src/api/c/blas.cpp:133
Type mismatch inputs

In function af::array af::matmulNT(const af::array &, const af::array &)
In file src/api/cpp/blas.cpp:28
Abort trap: 6
```

### CODE REVIEW

File: examples/alexnet.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            printf("Average Error at iteration(%d) : %lf\n", i + 1, af::mean<float>(af::abs(diff)));
+            printf("Predicted\n");
+            af_print(result.array());
+            printf("Expected\n");
+            af_print(out);
+            printf("\n\n");
+        }
+    }
+    return 0;
+}
```
Comment (pavanky) [inline examples/alexnet.cpp]: Can you send this in a different PR.

File: include/af/nn/Modules/Activations.hpp [hunk 2]
```diff
@@ -10,6 +10,7 @@
 
 #include <af/autograd/Variable.hpp>
 #include <af/nn/Modules/Module.hpp>
+#include <float.h>
```
Comment (pavanky) [inline include/af/nn/Modules/Activations.hpp]: Is this necessary ? If it is use `#include <cfloat>`

File: include/af/nn/Modules/Activations.hpp [hunk 3]
```diff
@@ -30,5 +31,12 @@ namespace af
 
             autograd::Variable forward(const autograd::Variable &input);
         };
+        class ReLU : public Module
```
Comment (pavanky) [inline include/af/nn/Modules/Activations.hpp]: Indentation

File: src/autograd/Functions.cpp [hunk 4]
```diff
@@ -55,6 +55,16 @@ namespace af {
             return Variable(result, {lhs, rhs}, grad_func);
         }
 
+        Variable operator >(const Variable &lhs, const Variable &rhs)
+        {
+            auto result = lhs.array() > rhs.array();
+            auto grad_func = [](std::vector<Variable> &inputs, const Variable &grad_output) {
+                inputs[0].addGrad(grad_output);
+                inputs[1].addGrad(grad_output);
```
Comment (pavanky) [inline src/autograd/Functions.cpp]: I don't think this is right. May be this function doesnt need to have a gradient?

File: src/autograd/Functions.cpp [hunk 5]
```diff
@@ -55,6 +55,16 @@ namespace af {
...
+            };
+            return Variable(result, {lhs, rhs}, grad_func);
```
Comment (pavanky) [inline src/autograd/Functions.cpp]: If you are removing grad_func, then `return Variable(result, false);` This tells the backward pass that this variable is not differentiable.

File: src/autograd/Functions.cpp [hunk 6]
```diff
@@ -147,6 +158,15 @@ namespace af {
...
         }
 
+        Variable relu(const Variable &input)
+        {
+	    auto result = max(input.array(), constant(0, input.array().dims()));
+	    auto grad_func = [](std::vector<Variable> &inputs, const Variable &grad_output) {
+	      inputs[0].addGrad(relu(grad_output) > 0);
+            };
+            return Variable(result, {input}, grad_func);
+        }
```
Comment (pavanky) [inline src/autograd/Functions.cpp]: I think this can be simplified like this:

```
Variable relu(const Variable &input)
          {
            auto mask = input > 0;
  	    auto result = mask.array() * input.array();
  	    auto grad_func = [](std::vector<Variable> &inputs, const Variable &grad_output) {
  	      inputs[0].addGrad(grad_output * inputs[1]);
              };
              return Variable(result, {input, mask}, grad_func);
          }
```

File: src/nn/Modules/Activations.cpp [hunk 7]
```diff
@@ -29,5 +29,12 @@ namespace af
         {
             return tanh(input);
         }
+
+        ReLU::ReLU() {}
+
+        Variable ReLU::forward(const Variable &input)
+        {
+  	    return relu(input);
+        }
```
Comment (pavanky) [inline src/nn/Modules/Activations.cpp]: :+1:

**Review by pavanky — CHANGES_REQUESTED** [review]

**Comment by pavanky** [comment]
@plavin Instead of having a `autograd::relu` function, it is better to implement an `autograd::max` function.

You can then call `autograd::max(var, 0)` from `nn::Relu::forward`

File: examples/alexnet.cpp (hunk 1 above)
Comment (pavanky) [reply examples/alexnet.cpp]: Or edit perceptron.cpp to use ReLU instead of sigmoid.

File: src/autograd/Functions.cpp (hunk 6 above)
Comment (plavin) [reply src/autograd/Functions.cpp]: What is inputs[1]?

Comment (pavanky) [reply src/autograd/Functions.cpp]: You are sending mask as second input for backward pass. You can use that instead of comparing inputs[0] with 0 again

Comment (pavanky) [reply src/autograd/Functions.cpp]: You are sending mask as second input for backward pass. You can use that instead of comparing inputs[0] with 0 again

**Review by pavanky — APPROVED** [review]
